src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,849 @@
+    cfg.MODEL.d_model = 768
+    cfg.MODEL.d_kv = 64
+    cfg.MODEL.d_ff = 3072
+    cfg.MODEL.num_layers = 12
+    cfg.MODEL.num_heads = 12
+    cfg.MODEL.dropout_prob = 0.1
+    cfg.MODEL.layer_norm_eps = 1E-6
+    cfg.MODEL.activation = 'relu'
+    cfg.MODEL.dtype = 'float32'
+    cfg.MODEL.layout = 'NT'
+    # initializer parameters
+    cfg.INITIALIZER = CN()
+    cfg.INITIALIZER.init_factor = 1.0
+    # other parameters
+    cfg.VERSION = 1
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_small(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 512
+    cfg.MODEL.d_ff = 2048
+    cfg.MODEL.num_layers = 6
+    cfg.MODEL.num_heads = 8
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_large(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_ff = 4096
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 16
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_3B(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_kv = 128
+    cfg.MODEL.d_ff = 16384
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 32
+    cfg.freeze()
+    return cfg
+
+
+@t5_cfg_reg.register()
+def google_t5_11B(): 
+    cfg = google_t5_base()
+    cfg.defrost()
+    cfg.MODEL.d_model = 1024
+    cfg.MODEL.d_kv = 128
+    cfg.MODEL.d_ff = 65536
+    cfg.MODEL.num_layers = 24
+    cfg.MODEL.num_heads = 128
+    cfg.freeze()
+    return cfg
+
+
+PRETRAINED_URL = {
+    'google_t5_small': {
+
+    }, 
+    'google_t5_base': {
+
+    }, 
+    'google_t5_large': {
+
+    }, 
+    'google_t5_3B': {
+
+    }, 
+    'google_t5_11B': {
+
+    }
+}
+
+
+# FILE_STATS = load_checksum_stats(os.path.join(get_model_zoo_checksum_dir(), 't5.txt'))
+
+
+@use_np
+class T5LayerNorm(HybridBlock): 
+    """
+    Layer normalization without bias and mean substraction
+    """
+    def __init__(self, d_model, epsilon=1E-6, init_factor=1.0, dtype='float32'): 
+        super().__init__()
+        self.weight = Parameter('layernorm_weight', shape=d_model, init=Constant(1.0 * init_factor), dtype=dtype)
+        self.variance_epsilon = epsilon
+
+    def forward(self, x): 
+        var = np.power(x.astype('float32'), 2).mean(-1, keepdims=True)
+        x = x * np.reciprocal(np.sqrt(var + self.variance_epsilon))
+        if self.weight.dtype == 'float16': 
+            x = x.astype('float16')
+        return self.weight.data() * x # TODO: .data()?
+
+
+@use_np
+class T5FeedForward(HybridBlock): 
+    """
+    Feed forward network supporting relu and gated-gelu
+    """
+    def __init__(
+        self, 
+        d_model, 
+        d_ff, 
+        dropout_prob=0.1, 
+        layer_norm_eps=1E-6, 
+        activation='relu', 
+        init_factor=1.0, 
+        dtype='float32'
+    ): 
+        super().__init__()
+        self._activation = activation
+
+        if activation == 'relu': 
```

Comment:
for this the conditional statements `if ... elif .. else` here, introducing an additional projection layer `wi_1` is sufficient

Reply:
That's a great point. I am now integrating this part into the `PositionwiseFFN` of layer.py.